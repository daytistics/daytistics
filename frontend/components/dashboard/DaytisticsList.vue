<template>
    <DashboardAddDaytisticDialog
        :open="isAddingDaytistic"
        @close="isAddingDaytistic = false"
    />

    <div class="bg-white shadow overflow-hidden sm:rounded-lg">
        <div class="px-4 py-5 sm:px-6 flex flex-row gap-3">
            <h3 class="text-lg leading-6 font-medium text-gray-900">
                Your Daytistics
            </h3>
            <button
                @click="isAddingDaytistic = true"
                class="text-secondary hover:text-secondary-dark flex items-center transition duration-150 ease-in-out"
            >
                <Plus class="w-5 h-5 mr-1" />
                <span class="text-sm font-medium">Add New</span>
            </button>
        </div>
        <div class="border-t border-gray-200">
            <div class="px-4 py-5 sm:p-6">
                <div class="flex items-center justify-between mb-4">
                    <input
                        type="text"
                        placeholder="Search date..."
                        class="shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded-md"
                    />
                    <button
                        class="ml-3 inline-flex items-center px-4 py-2 border border-transparent text-sm font-medium rounded-md text-white bg-secondary hover:bg-secondary-dark"
                    >
                        Search
                    </button>
                </div>
                <ul class="divide-y divide-gray-200">
                    <li
                        v-for="(daytistic, index) in daytistics"
                        :key="index"
                        class="py-4 cursor-pointer hover:bg-gray-50"
                    >
                        <NuxtLink :to="`/daytistics/${daytistic.id}`">
                            <div class="flex items-center space-x-4">
                                <Calendar class="h-6 w-6 text-gray-400" />
                                <div class="flex-1 min-w-0">
                                    <p
                                        class="text-sm font-medium text-gray-900 truncate"
                                    >
                                        {{
                                            new Date(
                                                daytistic.date
                                            ).toLocaleDateString()
                                        }}
                                    </p>
                                    <p class="text-sm text-gray-500 truncate">
                                        {{ daytistic.total_activities }}
                                        Activities /
                                        {{ daytistic.total_duration }}
                                    </p>
                                </div>
                            </div>
                        </NuxtLink>
                    </li>
                    <div class="flex justify-center">
                        <span
                            v-show="totalPages === 0"
                            class="text-lg"
                        >
                            We couldn't find any daytistics for you. Why not
                            <button
                                type="button"
                                class="text-secondary"
                            >
                                create
                            </button>
                            one?
                        </span>
                    </div>
                </ul>
                <div class="mt-4 flex items-center justify-between">
                    <button
                        :disabled="currentPage === 1"
                        @click="currentPage--"
                        class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"
                    >
                        Previous
                    </button>
                    <span
                        v-show="totalPages > 0"
                        class="text-sm text-gray-700"
                    >
                        Page {{ currentPage }} of {{ totalPages }}
                    </span>
                    <button
                        :disabled="currentPage === totalPages"
                        @click="currentPage++"
                        class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"
                    >
                        Next
                    </button>
                </div>
            </div>
        </div>
    </div>
</template>

<script lang="ts" setup>
import { Calendar, Plus } from 'lucide-vue-next';
import type { Daytistic } from '~/types/daytistics';
const daytistics = ref<Daytistic[] | []>([]);

const listDaytisticsAPI = useListDaytisticsAPI();

const currentPage = ref(1);
const totalPages = ref(1);
const isAddingDaytistic = ref(false);

watch(currentPage, async (newPage) => {
    if (totalPages.value > 0) {
        await listDaytisticsAPI.loadPage(newPage);
    }
});

onMounted(async () => {
    await listDaytisticsAPI.loadPage(currentPage.value);
});

function useListDaytisticsAPI() {
    const { $api } = useNuxtApp();

    const list = async (page: number) => {
        return $api(`/api/daytistics/list?page=${page}`, {
            server: false,
            method: 'GET',
        });
    };

    const loadPage = async (page: number) => {
        const response: any = await list(page);
        daytistics.value = response.items;
        totalPages.value = Math.ceil(response.count / 5);
    };
    return {
        loadPage,
    };
}
</script>
