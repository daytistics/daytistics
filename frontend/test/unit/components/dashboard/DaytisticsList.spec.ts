import { it, describe, expect, vi, beforeEach } from 'vitest';
import { mockNuxtImport, mountSuspended } from '@nuxt/test-utils/runtime';
import { flushPromises } from '@vue/test-utils';
import DaytisticsList from '~/components/dashboard/DaytisticsList.vue';
import type { Daytistic } from '~/types/daytistics';
import { findByAttribute, findByText } from '~/test/test-utils/selectors';
import { generateDaytistics } from '~/test/test-utils/mocks';

const DAYTISTICS_COUNT = 9;
const mockListDaytistics = vi.fn();

let daytistics: undefined | Daytistic[] = undefined;

vi.mock('~/composables/useEndpoints', () => ({
    default: vi.fn(() => ({
        listDaytistics: mockListDaytistics,
    })),
}));

describe('DaytisticsList', () => {
    beforeEach(() => {
        vi.clearAllMocks();
        daytistics = generateDaytistics(DAYTISTICS_COUNT);

        mockListDaytistics.mockResolvedValue({
            count: DAYTISTICS_COUNT,
            items: daytistics,
        });
    });

    it('should render the list', async () => {
        const wrapper = await mountSuspended(DaytisticsList);
        await flushPromises();

        expect(mockListDaytistics).toHaveBeenCalledTimes(1);
        expect(wrapper.text()).toContain('Your Daytistics');
        expect(wrapper.text()).toContain(`1 of 2`);
        expect(wrapper.text()).toContain('Add New');
        expect(wrapper.text()).toContain('Next');
        expect(wrapper.text()).toContain('Previous');

        for (const daytistic of daytistics?.slice(0, 5)!) {
            const date = new Date(daytistic.date).toLocaleDateString();
            expect(wrapper.text()).toContain(date);
        }
    });

    it('matches snapshot', async () => {
        const wrapper = await mountSuspended(DaytisticsList);
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("switches to the next page when the 'next' button is clicked", async () => {
        const wrapper = await mountSuspended(DaytisticsList);
        await flushPromises();

        const nextButton = findByText(wrapper, 'button', /Next/).at(0);
        await nextButton!.trigger('click');

        await flushPromises();

        expect(mockListDaytistics).toHaveBeenCalledTimes(2);
        expect(wrapper.text()).toContain('2 of 2');

        for (const daytistic of daytistics?.slice(5, 10)!) {
            const date = new Date(daytistic.date).toLocaleDateString();
            expect(wrapper.text()).toContain(date);
        }
    });

    it("switches to the previous page when the 'previous' button is clicked", async () => {
        const wrapper = await mountSuspended(DaytisticsList);

        await flushPromises();

        const nextButton = findByText(wrapper, 'button', /Next/).at(0);
        await nextButton!.trigger('click');

        await flushPromises();

        const previousButton = findByText(wrapper, 'button', /Previous/).at(0);
        await previousButton!.trigger('click');

        await flushPromises();

        expect(mockListDaytistics).toHaveBeenCalledTimes(3);
        expect(wrapper.text()).toContain('1 of 2');

        for (const daytistic of daytistics?.slice(0, 5)!) {
            const date = new Date(daytistic.date).toLocaleDateString();
            expect(wrapper.text()).toContain(date);
        }
    });

    it("opens the 'CreateDaytisticDialog' when the 'Add Daytistic' button is clicked", async () => {
        const wrapper = await mountSuspended(DaytisticsList);
        await flushPromises();

        const addButton = findByText(wrapper, 'button', /Add New/).at(0);
        await addButton!.trigger('click');

        wrapper.vm.$nextTick();

        expect(findByAttribute(wrapper, 'span', 'data-dialog-open-test', 'true').length).toBe(1);
    });
});
