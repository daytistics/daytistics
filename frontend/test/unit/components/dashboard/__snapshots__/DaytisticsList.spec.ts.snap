// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DaytisticsList > matches snapshot 1`] = `
"<!-- Dialog to create a new Daytistic - the span is used for testing purposes -->
<span aria-hidden="true" data-dialog-open-test="false"></span>
<!--teleport start-->
<!--teleport end-->
<div hidden="" style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; border-width: 0px; display: none;"></div>
<!-- Actual list -->
<div class="bg-white shadow overflow-hidden sm:rounded-lg">
  <div class="px-4 py-5 sm:px-6 flex flex-row gap-3 justify-between items-center">
    <h3 class="text-lg leading-6 font-medium text-gray-900">Your Daytistics</h3><button class="text-secondary hover:text-tertiary flex items-center transition duration-150 ease-in-out"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-plus-icon w-6 md:w-5 h-auto md:mr-1">
        <path d="M5 12h14"></path>
        <path d="M12 5v14"></path>
      </svg><span class="text-sm font-medium hidden md:block">Add New</span></button>
  </div>
  <div class="border-t border-gray-200">
    <div class="px-4 py-5 sm:p-6">
      <ul class="divide-y divide-day-gray-medium">
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/1">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/2/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/2">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/3/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/3">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/4/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/4">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/5/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/5">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/6/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/6">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/7/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/7">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/8/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/8">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/9/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <li class="py-4 cursor-pointer hover:bg-day-gray-light"><a href="/dashboard/daytistics/9">
            <div class="flex items-center space-x-4 px-3"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-calendar-icon h-6 w-6 text-day-gray-dark">
                <path d="M8 2v4"></path>
                <path d="M16 2v4"></path>
                <rect width="18" height="18" x="3" y="4" rx="2"></rect>
                <path d="M3 10h18"></path>
              </svg>
              <div class="flex-1 min-w-0">
                <p class="text-sm font-medium text-dark truncate">1/10/1970</p>
                <p class="text-sm text-day-gray-dark truncate">3 Activities / 3h 0m</p>
              </div>
            </div>
          </a></li>
        <div class="flex justify-center"><span class="text-md text-center" style="display: none;"> We couldn't find any Daytistics for you. Why not <button type="button" class="text-secondary"> create </button> one? </span></div>
      </ul>
      <div class="mt-4 flex items-center justify-between"><button disabled="" class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"><span class="hidden md:block">Previous</span><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-left-icon md:hidden">
            <path d="m15 18-6-6 6-6"></path>
          </svg></button><span class="text-sm text-gray-700"><span class="hidden md:inline-block">Page</span> 1 of 2</span><button class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"><span class="hidden md:block">Next</span><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-right-icon md:hidden">
            <path d="m9 18 6-6-6-6"></path>
          </svg></button></div>
    </div>
  </div>
</div>"
`;
